lib/base/string.cpp

```diff
@@ -333,11 +333,21 @@ String icinga::operator+(const String& lhs, const char *rhs)
 	return lhs.GetData() + rhs;
 }
 
+String icinga::operator+(String&& lhs, const char *rhs)
+{
+	return std::move(lhs.GetData()) + rhs;
+}
+
 String icinga::operator+(const char *lhs, const String& rhs)
 {
 	return lhs + rhs.GetData();
 }
 
+String icinga::operator+(const char *lhs, String&& rhs)
+{
+	return lhs + std::move(rhs.GetData());
+}
```

Comment:
How did you get here? Like how did you get to the conclusion that this will be a useful optimization?

(Also, I really wonder if there's actually an implementation of `std::string` that allows prepending without reallocation (unless you prepend the empty string), but well, a corresponding overload of `operator+` is provided for `std::string`.)

Reply:
> How did you get here? Like how did you get to the conclusion that this will be a useful optimization?

Checkout 27e1850381220ce3f997503bf0cae3dfc7cd1896 and Cmd+LClick the + sign in https://github.com/Icinga/icinga2/blob/27e1850381220ce3f997503bf0cae3dfc7cd1896/lib/base/json.cpp#L199

CLion will open `String icinga::operator+(const String&,const char*)` which **copies** the string.

> (Also, I really wonder if there's actually an implementation of `std::string` that allows prepending without reallocation

As long as there's enough free capacity, there's no reason to reallocate.

> a corresponding overload of `operator+` is provided for `std::string`.)

Exactly.